File: backend/app/models/analytics.py
from pydantic import BaseModel
from typing import List, Optional
from uuid import UUID
from datetime import date

# Forecast Models
class ForecastRequest(BaseModel):
    product_id: UUID
    warehouse_id: UUID

class ForecastResponse(BaseModel):
    product_id: UUID
    warehouse_id: UUID
    forecast_daily_demand: float
    method: str
    historical_period_used_days: int

# Risk Models
class RiskScoreResponse(BaseModel):
    batch_id: UUID
    product_id: UUID
    warehouse_id: UUID
    risk_score: float
    risk_level: str
    days_to_expiry: int
    expected_demand_before_expiry: float
    potential_excess: float

# Redistribution Models
class RedistributionRecommendation(BaseModel):
    product_id: UUID
    source_warehouse_id: UUID
    destination_warehouse_id: UUID
    recommended_quantity: int
    source_risk_level: str
    destination_demand: float
    distance_km: float
    reason: str

class RedistributionResponse(BaseModel):
    recommendations: List[RedistributionRecommendation]

# Simulation Models
class WhatIfSimulationRequest(BaseModel):
    recommendation: RedistributionRecommendation

class WhatIfSimulationResponse(BaseModel):
    waste_before_transfer: float
    waste_after_transfer: float
    waste_avoided: float

# Gemini Explanation Models
class ExplainRequest(BaseModel):
    product_name: str
    warehouse_name: str
    quantity: Optional[float] = None
    days_to_expiry: int
    risk_score: float
    risk_level: str
    potential_excess: Optional[float] = None
    forecast_daily_demand: Optional[float] = None
    recommended_transfer: Optional[int] = None
    destination_warehouse: Optional[str] = None
    destination_demand: Optional[float] = None
    transfer_distance_km: Optional[float] = None

class ExplainResponse(BaseModel):
    explanation: str

File: frontend/src/types/analytics.ts
export interface Warehouse {
  id: string;
  name: string;
  location: string;
  latitude?: number;
  longitude?: number;
  capacity: number;
  created_at: string;
}

export interface Product {
  id: string;
  name: string;
  category: string;
  unit: string;
  created_at: string;
}

export interface InventoryBatch {
  id: string;
  product_id: string;
  warehouse_id: string;
  quantity: number;
  received_date: string;
  expiry_date: string;
  created_at: string;
}

export interface RiskScoreResponse {
  batch_id: string;
  product_id: string;
  warehouse_id: string;
  risk_score: number;
  risk_level: 'LOW' | 'MEDIUM' | 'HIGH' | 'CRITICAL';
  days_to_expiry: number;
  expected_demand_before_expiry: number;
  potential_excess: number;
}

export interface ForecastResponse {
  product_id: string;
  warehouse_id: string;
  forecast_daily_demand: number;
  method: string;
  historical_period_used_days: number;
}

export interface RedistributionRecommendation {
  product_id: string;
  source_warehouse_id: string;
  destination_warehouse_id: string;
  recommended_quantity: number;
  source_risk_level: string;
  destination_demand: number;
  distance_km: number;
  reason: string;
}

export interface RedistributionResponse {
  recommendations: RedistributionRecommendation[];
}

export interface WhatIfSimulationRequest {
  recommendation: RedistributionRecommendation;
}

export interface WhatIfSimulationResponse {
  waste_before_transfer: number;
  waste_after_transfer: number;
  waste_avoided: number;
}

// Gemini Explanation
export interface ExplainInventoryRequest {
  product_name: string;
  warehouse_name: string;
  quantity?: number;
  days_to_expiry: number;
  risk_score: number;
  risk_level: string;
  potential_excess?: number;
  forecast_daily_demand?: number;
  recommended_transfer?: number;
  destination_warehouse?: string;
  destination_demand?: number;
  transfer_distance_km?: number;
}

export interface ExplainInventoryResponse {
  explanation: string;
}
